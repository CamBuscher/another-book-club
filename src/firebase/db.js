import { db } from './firebase';

export const doCreateUser = (id, username, email) => 
  db.ref(`users/${id}`).set({
    username,
    email,
    id
  });

export const updateUserBookClubs = (user, newClub) => {
  const existingClubs = user.bookClubs || [];
  const updatedUser = Object.assign({...user}, {bookClubs: [...existingClubs, newClub]})
  db.ref(`users/${user.id}/bookClubs/${newClub.clubName}`).set({...newClub});
};

export const doCreateBookClub = (id, clubName, member) =>
  db.ref(`bookClubs/${id}`).set({
    clubName,
    members: [...member],
    id
  });

export const onceGetUsers = () => 
  db.ref('users').once('value');