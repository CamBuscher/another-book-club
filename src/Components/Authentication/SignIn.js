import React, { Component } from 'react';
import { withRouter } from 'react-router-dom';
import PropTypes from 'prop-types';
import { SignUpLink } from './SignUp';
import { auth } from '../../firebase';
import * as routes from '../../constants/routes';

const SignInPage = ({ history }) =>
  <div>
    <h1>Sign In</h1>
    <SignInForm history={history} />
    <SignUpLink />
  </div>;

const byPropKey = (propertyName, value) => () => ({
  [propertyName]: value
});

export const INITIAL_STATE = {
  email: '',
  password: '',
  error: null
};

class SignInForm extends Component {
  constructor(props) {
    super(props);

    this.state = { ...INITIAL_STATE };
  }

  onSubmit = (event) => {
    const {
      email,
      password
    } = this.state;

    const {
      history
    } = this.props;

    auth.doSignInWithEmailAndPassword(email, password)
      .then(() => {
        this.setState(() => ({ ...INITIAL_STATE }));
        history.push(routes.ACCOUNT);
      })
      .catch(error => {
        this.setState(byPropKey('error', error));
      });
    event.preventDefault();
  }

  render() {
    const {
      email,
      password,
      error
    } = this.state;

    const isInvalid =
      password === '' ||
      email === '';

    return (
      <form onSubmit={this.onSubmit} className='signIn authentication'>
        <input
          value={email}
          onChange={event => this.setState(byPropKey('email', event.target.value))}
          type="text"
          placeholder="Email Address"
        />
        <input
          value={password}
          onChange={event => this.setState(byPropKey('password', event.target.value))}
          type="password"
          placeholder="Password"
        />
        <button disabled={isInvalid} type="submit">
          Sign In
        </button>

        { error && <p>{error.message}</p> }
      </form>
    );
  }
}

SignInPage.propTypes = {
  history: PropTypes.object
};

SignInForm.propTypes = {
  history: PropTypes.object
};

export default withRouter(SignInPage);

export {
  SignInForm
};