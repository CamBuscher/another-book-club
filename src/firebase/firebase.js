import firebase from 'firebase/app';
import { firebaseKEY } from '../Helpers/APIkey';
import 'firebase/auth';
import 'firebase/database'

const config = {
  apiKey: `${firebaseKEY}`,
  authDomain: "fir-tutorial-db803.firebaseapp.com",
  databaseURL: "https://fir-tutorial-db803.firebaseio.com",
  projectId: "fir-tutorial-db803",
  storageBucket: "fir-tutorial-db803.appspot.com",
  messagingSenderId: "318822888465"
};

if (!firebase.apps.length) {
  firebase.initializeApp(config);
} 

const db = firebase.database()
const auth = firebase.auth();

export {
  db,
  auth,
};