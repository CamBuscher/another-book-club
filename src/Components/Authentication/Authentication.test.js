import {SignInForm, INITIAL_STATE} from './SignIn';
import SignOutButton from './SignOut';
import {SignUpForm} from './SignUp';
import React from 'react';
import { shallow, mount } from 'enzyme';
import { auth } from '../../firebase';

describe('signInForm', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<SignInForm />)
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should start with the correct initial state', () => {
    expect(wrapper.state()).toEqual(INITIAL_STATE);
  });

  it('should update state when the input field receives input', () => {
    const input = wrapper.find('input').first();

    input.simulate('change', {target: {
      value: 'hello'
    }});

    expect(wrapper.state().email).toEqual('hello');
  });

  it('should update state when the second input field receives input', () => {
    const input = wrapper.find('input').at(1);

    input.simulate('change', {
      target: {
        value: 'hello'
      }
    });

    expect(wrapper.state().password).toEqual('hello');
  });
  
//   it('should call auth.doSignIn function on submit', async () => {
//     wrapper = mount(<SignInForm />);

//     wrapper.find('input').at(1).simulate('change', {
//       target: {
//         value: ''
//       }
//     });

//     wrapper.find('input').at(0).simulate('change', {
//       target: {
//         value: 'hello'
//       }
//     });

//     wrapper.find('form').simulate('submit', {
//       preventDefault: jest.fn()
//     });

//     await expect(auth.doSignInWithEmailAndPassword).toHaveBeenCalled();
//   });

//   it('should reset initial state on submit', async () => {
//     wrapper = mount(<SignInForm />);

//     wrapper.find('input').at(1).simulate('change', {
//       target: {
//         value: ''
//       }
//     });

//     wrapper.find('input').at(0).simulate('change', {
//       target: {
//         value: 'hello'
//       }
//     });

//     wrapper.find('form').simulate('submit', {
//       preventDefault: jest.fn()
//     });

//     await expect(wrapper.state()).toEqual(INITIAL_STATE);
//   });
});

describe('Sign Out', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<SignOutButton />);
  });
  
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call auth.doSignOut', () => {
    wrapper.simulate('click')

    const auth = {
      doSignOut: jest.fn()
    };

    expect(auth.doSignOut).toHaveBeenCalled();
  });
});

describe('Sign Up', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<SignUpForm />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })
});