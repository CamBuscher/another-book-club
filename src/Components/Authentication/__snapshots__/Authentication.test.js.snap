// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sign Out should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <withRouter() />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "render": [Function],
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "render": [Function],
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`Sign Up should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignUpForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onChange={[Function]}
          placeholder="Full Name"
          type="text"
          value=""
/>,
        <input
          onChange={[Function]}
          placeholder="Email Address"
          type="text"
          value=""
/>,
        <input
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
/>,
        <input
          onChange={[Function]}
          placeholder="Confirm Password"
          type="password"
          value=""
/>,
        <button
          disabled={true}
          type="submit"
>
          Sign Up
</button>,
        null,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Full Name",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Email Address",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Confirm Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign Up",
          "disabled": true,
          "type": "submit",
        },
        "ref": null,
        "rendered": "Sign Up",
        "type": "button",
      },
      null,
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onChange={[Function]}
            placeholder="Full Name"
            type="text"
            value=""
/>,
          <input
            onChange={[Function]}
            placeholder="Email Address"
            type="text"
            value=""
/>,
          <input
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
/>,
          <input
            onChange={[Function]}
            placeholder="Confirm Password"
            type="password"
            value=""
/>,
          <button
            disabled={true}
            type="submit"
>
            Sign Up
</button>,
          null,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Full Name",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Email Address",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Confirm Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign Up",
            "disabled": true,
            "type": "submit",
          },
          "ref": null,
          "rendered": "Sign Up",
          "type": "button",
        },
        null,
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;

exports[`signInForm should match snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <SignInForm />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <input
          onChange={[Function]}
          placeholder="Email Address"
          type="text"
          value=""
/>,
        <input
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
/>,
        <button
          disabled={true}
          type="submit"
>
          Sign In
</button>,
        null,
      ],
      "className": "signIn",
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Email Address",
          "type": "text",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "onChange": [Function],
          "placeholder": "Password",
          "type": "password",
          "value": "",
        },
        "ref": null,
        "rendered": null,
        "type": "input",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Sign In",
          "disabled": true,
          "type": "submit",
        },
        "ref": null,
        "rendered": "Sign In",
        "type": "button",
      },
      null,
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <input
            onChange={[Function]}
            placeholder="Email Address"
            type="text"
            value=""
/>,
          <input
            onChange={[Function]}
            placeholder="Password"
            type="password"
            value=""
/>,
          <button
            disabled={true}
            type="submit"
>
            Sign In
</button>,
          null,
        ],
        "className": "signIn",
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Email Address",
            "type": "text",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "onChange": [Function],
            "placeholder": "Password",
            "type": "password",
            "value": "",
          },
          "ref": null,
          "rendered": null,
          "type": "input",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Sign In",
            "disabled": true,
            "type": "submit",
          },
          "ref": null,
          "rendered": "Sign In",
          "type": "button",
        },
        null,
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
